lib/travis/api/v3/queries/log.rb

```diff
@@ -0,0 +1,49 @@
+module Travis::API::V3
+  class Queries::Log < Query
+    require 'net/http'
+    require 'uri'
+
+    FORMAT = "Log removed by %s at %s"
+
+    def find(job)
+      #check for the log in the Logs DB
+      log = Models::Log.find_by_job_id(job.id)
+
+      raise EntityMissing, 'log not found'.freeze if log.nil?
+      #if the log has been archived, go to s3
+      if log.archived_at
+        ## if it's not there then fetch it from S3, and return it wrapped as a
+        ## compatible log_parts object with a hard coded number (log_parts have a number)
+        ## and a single log_part that contains all the log content
+        archived_log_path = archive_url("/jobs/#{job.id}/log.txt")
+        content = Net::HTTP.get(URI.parse(archived_log_path))
+        log_part = Models::LogPart.new(log_id: log.id, content: content, number: 0, created_at: log.created_at)
```

Comment:
There's new behaviour here. Instead of creating a Log object with many Log Parts, the txt of the archive is put into a single Log Part. This saves on cpu usage and is similar to how archived logs were previously handled. See Sven for more details.